diff --git a/scripts/boot_patch.sh b/scripts/boot_patch.sh
index 068097c38e4574653a04fcf188bbf41925102802..646b16511d3d87dddfce00106520a1611e04f2f5 100644
--- a/scripts/boot_patch.sh
+++ b/scripts/boot_patch.sh
@@ -136,6 +136,22 @@ case $((STATUS & 3)) in
     ;;
 esac
 
+# Check vivo boot image
+./magiskboot cpio ./ramdisk.cpio exists /init.vivo.rc 2>/dev/null
+if [ $? -ne 0 ]; then
+    VIVO=0
+else
+    ui_print "- init.vivo.rc detected"
+    VIVO=1
+    ./magiskboot cpio ./ramdisk.cpio exists /fstab.qcom 2>/dev/null
+    if [ $? -ne 0 ]; then
+        ui_print "! Coundn't find fstab.qcom"
+        QCOM=0
+    else
+        QCOM=1
+    fi
+fi
+
 # Work around custom legacy Sony /init -> /(s)bin/init_sony : /init.real setup
 INIT=init
 if [ $((STATUS & 4)) -ne 0 ]; then
@@ -146,6 +162,32 @@ fi
 # Ramdisk Patches
 ##################
 
+# Patch ramdisk for vivo
+if [ $VIVO -ne 0 ]; then
+    ui_print "- Patching ramdisk for vivo devices"
+    mkdir vivo
+    ./magiskboot cpio ./ramdisk.cpio \
+    "extract /init.vivo.rc ./vivo/"
+    ./magiskboot hexpatch ./vivo/init.vivo.rc \
+    237669766F5F6461656D6F6E206164646564206279207A68616E676D696E676B61690A73657276696365207669766F5F6461656D6F6E202F73797374656D2F62696E2F7669766F5F6461656D6F6E0A20202020636C617373206D61696E \
+    2348692C7669766F2C7669766F5F6461656D6F6E2069732064697361626C656421210A234F4B2C746869732073657276696365207669766F5F6461656D6F6E2069732064697361626C65642121210A23446F6E27742073746172742121
+    ./magiskboot cpio ./ramdisk.cpio \
+    "rm /init.vivo.rc"
+    "add 0750 /init.vivo.rc ./vivo/init.vivo.rc"
+    if [ $KEEPFORCEENCRYPT -ne false ];then
+        if [ $QCOM -ne 0 ];then
+            ui_print "- Patching fstab to fix encrypt"
+            ./magiskboot hexpatch ./vivo/fstab.qcom \
+            636865636B2C66F726365656E63727970743D666F6F6F746572 \
+            636865636B2656E656E63727970743D666F6F6F746572202020
+            ./magiskboot cpio ./ramdisk.cpio \
+            "rm /fstab.qcom"
+            "add 0640 /fstab.qcom ./vivo/fstab.qcom"
+        fi
+    fi
+    rm -rf ./vivo
+fi
+
 ui_print "- Patching ramdisk"
 
 echo "KEEPVERITY=$KEEPVERITY" > config
@@ -207,6 +249,14 @@ if [ -f kernel ]; then
   # Before: [mov w2, #-221]   (-__NR_execve)
   # After:  [mov w2, #-32768]
   ./magiskboot hexpatch kernel 821B8012 E2FF8F12
+  
+  # Patch vivo kernel (mount system)
+  if [ $VIVO -ne 0 ];then
+      ui_print "- Patching kernel to fix mount"  
+      ./magiskboot hexpatch kernel \
+      0092CFC2C9CDDDDA00 \
+      0092CFC2C9CEC0DB00 # Add by wuxianlin
+  fi
 
   # Force kernel to load rootfs
   # skip_initramfs -> want_initramfs
